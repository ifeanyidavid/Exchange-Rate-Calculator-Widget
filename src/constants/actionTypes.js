export const FETCH_RATES_START = "FETCH_RATES_START";
export const FETCH_RATES_ERROR = "FETCH_RATES_ERROR";
export const FETCH_RATES_DONE = "FETCH_RATES_DONE";
export const UPDATE_SOURCE_CURRENCY = "UPDATE_SOURCE_CURRENCY";
export const UPDATE_DESTINATION_CURRENCY = "UPDATE_DESTINATION_CURRENCY";
export const EXCHANGE_CURRENCY = "EXCHANGE_CURRENCY";
