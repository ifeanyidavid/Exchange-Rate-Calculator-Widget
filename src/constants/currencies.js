export const currencies = {
  EUR: { id: 0, value: "EUR" },
  GBP: { id: 1, value: "GBP" },
  USD: { id: 2, value: "USD" }
};
