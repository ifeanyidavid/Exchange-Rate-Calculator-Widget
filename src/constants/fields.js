export const fieldNames = {
  sourceCurrency: "sourceCurrency",
  destinationCurrency: "destinationCurrency",
  sourceAmount: "sourceAmount",
  destinationAmount: "destinationAmount",
  balance: "balance",
  symbol: "symbol",
  description: "description",
  currency: "currency",
  id: "id",
  error: "error",
  loading: "loading",
  rates: "rates",
};
