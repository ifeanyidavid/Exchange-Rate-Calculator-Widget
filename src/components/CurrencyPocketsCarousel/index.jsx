import React from "react";
import Slider from "react-slick";
import PropTypes from "prop-types";
import styled from "styled-components";
import "slick-carousel/slick/slick.css";
import "slick-carousel/slick/slick-theme.css";
import { currencies } from "../../constants/currencies";

const SliderContainer = styled.div`
  color: #fff;
  min-height: 300px;
  max-height: 100%;
  cursor: move;
  display: flex !important;
  text-align: center !important;
  flex-direction: column !important;
  justify-content: center !important;
`;

const BlockText = styled.span`
  font-size: ${props => props.fontSize};
  display: block;
  weight: ${props => props.weight};
`;

function CurrencyPocketsCarousel({ updateSourceCurrency, pockets, currency }) {
  const settings = {
    dots: true,
    infinite: true,
    speed: 500,
    slidesToShow: 1,
    slidesToScroll: 1,
    initialSlide: currencies[currency].id,
    swipeToSlide: true,
    responsive: [
      {
        breakpoint: 1024,
        settings: {
          slidesToShow: 1,
          slidesToScroll: 1,
          infinite: true,
          dots: true
        }
      },
      {
        breakpoint: 600,
        settings: {
          slidesToShow: 1,
          slidesToScroll: 1,
          initialSlide: 2
        }
      },
      {
        breakpoint: 480,
        settings: {
          slidesToShow: 1,
          slidesToScroll: 1
        }
      }
    ]
  };

  const handleSlide = index => {
    updateSourceCurrency(index);
  };

  return (
    <>
      <Slider afterChange={handleSlide} {...settings}>
        {pockets.map(pocket => (
          <SliderContainer key={pocket.id}>
            <div>
              <BlockText weight="bold" fontSize="32px">
                {`${pocket.symbol} ${parseFloat(pocket.balance).toFixed(2)}`}
              </BlockText>
              <BlockText weight="normal" fontSize="16px">
                {`${pocket.currency} - ${pocket.description}`}
              </BlockText>
            </div>
          </SliderContainer>
        ))}
      </Slider>
    </>
  );
}

export default CurrencyPocketsCarousel;

CurrencyPocketsCarousel.propTypes = {
  updateSourceCurrency: PropTypes.func.isRequired,
  pockets: PropTypes.array.isRequired,
  currency: PropTypes.string.isRequired
};
