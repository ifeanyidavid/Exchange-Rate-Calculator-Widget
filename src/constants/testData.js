import { currencies } from "./currencies";
import { fieldNames } from "./fields";

export const testData = {
  rates: [
    {
      rates: { USD: 1.1075, GBP: 0.84195 },
      base: "EUR",
      date: "2019-12-09"
    },
    {
      rates: { EUR: 1.1877189857, USD: 1.3153987766 },
      base: "GBP",
      date: "2019-12-09"
    },
    {
      rates: { EUR: 0.9029345372, GBP: 0.7602257336 },
      base: "USD",
      date: "2019-12-09"
    }
  ],
  pocketsReducer: {
    transactionHistory: [
      {
        id: 0,
        description: `Exchanged to EUR`,
        time: new Date().toLocaleTimeString("en-GB"),
        sourceAmount: 23,
        destinationAmount: 42,
        sourceCurrency: "USD",
        destinationCurrency: "EUR"
      }
    ],
    pockets: {
      [currencies.EUR.value]: {
        [fieldNames.id]: currencies.EUR.id,
        [fieldNames.currency]: currencies.EUR.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "€",
        [fieldNames.description]: "Euro"
      },
      [currencies.GBP.value]: {
        [fieldNames.id]: currencies.GBP.id,
        [fieldNames.currency]: currencies.GBP.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "£",
        [fieldNames.description]: "British Pound"
      },
      [currencies.USD.value]: {
        [fieldNames.id]: currencies.USD.id,
        [fieldNames.currency]: currencies.USD.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "$",
        [fieldNames.description]: "American Dollar"
      }
    }
  },
  sourceAmount: 23,
  destinationAmount: 54,
  sourceCurrency: "USD",
  destinationCurrency: "EUR",
  error: new Error("Network disconnected"),
  slideIndex: 1,
  pocketValues: [
    {
      [fieldNames.id]: currencies.EUR.id,
      [fieldNames.currency]: currencies.EUR.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "€",
      [fieldNames.description]: "Euro"
    },
    {
      [fieldNames.id]: currencies.GBP.id,
      [fieldNames.currency]: currencies.GBP.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "£",
      [fieldNames.description]: "British Pound"
    },
    {
      [fieldNames.id]: currencies.USD.id,
      [fieldNames.currency]: currencies.USD.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "$",
      [fieldNames.description]: "American Dollar"
    }
  ]
};
