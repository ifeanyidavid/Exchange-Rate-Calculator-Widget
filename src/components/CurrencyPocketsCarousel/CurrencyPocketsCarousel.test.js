import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
import toJson from "enzyme-to-json";
import CurrencyPocketsCarousel from ".";
import { fieldNames } from "../../constants/fields";
import { currencies } from "../../constants/currencies";

configure({ adapter: new Adapter() });

describe("CurrencyPocketsCarousel component", () => {
  it("Renders correctly", () => {
    const pockets = [
      {
        [fieldNames.id]: currencies.EUR.id,
        [fieldNames.currency]: currencies.EUR.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "€",
        [fieldNames.description]: "Euro"
      },
      {
        [fieldNames.id]: currencies.GBP.id,
        [fieldNames.currency]: currencies.GBP.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "£",
        [fieldNames.description]: "British Pound"
      },
      {
        [fieldNames.id]: currencies.USD.id,
        [fieldNames.currency]: currencies.USD.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "$",
        [fieldNames.description]: "American Dollar"
      }
    ];

    const props = {
      updateSourceCurrency: jest.fn(),
      pockets,
      currency: "USD"
    };

    // wrapper instance around rendered output
    const wrapper = shallow(<CurrencyPocketsCarousel {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
