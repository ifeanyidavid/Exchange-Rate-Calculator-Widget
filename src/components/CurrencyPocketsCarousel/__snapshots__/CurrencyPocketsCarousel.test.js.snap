// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyPocketsCarousel component Renders correctly 1`] = `
<Fragment>
  <Slider
    afterChange={[Function]}
    dots={true}
    infinite={true}
    initialSlide={2}
    responsive={
      Array [
        Object {
          "breakpoint": 1024,
          "settings": Object {
            "dots": true,
            "infinite": true,
            "slidesToScroll": 1,
            "slidesToShow": 1,
          },
        },
        Object {
          "breakpoint": 600,
          "settings": Object {
            "initialSlide": 2,
            "slidesToScroll": 1,
            "slidesToShow": 1,
          },
        },
        Object {
          "breakpoint": 480,
          "settings": Object {
            "slidesToScroll": 1,
            "slidesToShow": 1,
          },
        },
      ]
    }
    slidesToScroll={1}
    slidesToShow={1}
    speed={500}
    swipeToSlide={true}
  >
    <styled.div
      key="0"
    >
      <div>
        <styled.span
          fontSize="32px"
          weight="bold"
        >
          € 300.00
        </styled.span>
        <styled.span
          fontSize="16px"
          weight="normal"
        >
          EUR - Euro
        </styled.span>
      </div>
    </styled.div>
    <styled.div
      key="1"
    >
      <div>
        <styled.span
          fontSize="32px"
          weight="bold"
        >
          £ 300.00
        </styled.span>
        <styled.span
          fontSize="16px"
          weight="normal"
        >
          GBP - British Pound
        </styled.span>
      </div>
    </styled.div>
    <styled.div
      key="2"
    >
      <div>
        <styled.span
          fontSize="32px"
          weight="bold"
        >
          $ 300.00
        </styled.span>
        <styled.span
          fontSize="16px"
          weight="normal"
        >
          USD - American Dollar
        </styled.span>
      </div>
    </styled.div>
  </Slider>
</Fragment>
`;
